Nette 3

masicek opened the pull request.

NoxZet committed:
Nette 3.0 v composer.json

NoxZet committed:
peldax/RabbitMq fork (Nette 3.0) v composer.json

NoxZet committed:
Replace deprecated setInject with addTag TAG_INJECT

NoxZet committed:
Use Kdyby EventManager for console termination listener

thorewi commented on src/DI/BackgroundQueueExtension.php:
```diff
@@ -102,8 +103,8 @@ public function afterCompile(ClassType $class)
-	$this->getByType(\Symfony\Component\EventDispatcher\EventDispatcherInterface::class)
-		->addListener(\Symfony\Component\Console\ConsoleEvents::TERMINATE, function () use ($service) {
+	$this->getByType(\Kdyby\Events\EventManager::class)
+		->addEventListener(\Symfony\Component\Console\ConsoleEvents::TERMINATE, function () use ($service) {
```
@NoxZet proc se delalo toto?

NoxZet replied:
@thorewi už si úplně nepamatuji, předpokládám, že mi to s novýma knihovnama nějak přestalo fungovat

thorewi replied:
@NoxZet tak to jeste jednou testni s tim puvodnim a dej mi vedet co tam byl za problem...

thorewi replied:
a testni to na 2.4 i na 3

NoxZet replied:
@thorewi chceme Tedy nechceme vůbec používat Kdyby. Problém byl že Kdyby má od nějaké verze nějakou vlastní implementaci EventDispatcherInterface, které pak naimplementovanou metodu addListener, proto to někde kde se používá Kdyby to přestalo fungovat. Zřejmě tedy pokusit se odstranit Kdyby na projektu a tady to vrátit do předchozího stavu.

thorewi replied:
@NoxZet jo zkusme to udelat bez zavislosti na kdyby, to mi dava smysl...

NoxZet replied:
@thorewi mělo by to být ready na merge, vydal bych to tedy jako novou major verzi kvůli rozbití zpětné kompatibility v Nette 3

thorewi replied:
ok ale urcite se mi jeste nelibi zavislost na nette 3.0... vyjmenujme konkretni baliky, ktere to potrebuje...

thorewi replied:
a co presne rozbiji tu zpetnou kompatibilitu s nette 2.4?

NoxZet replied:
Jak to tak zkoumám, myslím že kód samotný by měl fungovat na obou. V composer.json ale třeba `kdyby/rabbitmq` 1.1/1.2 určitě nebude fungovat s Nette 3 a naopak `adt/rabbitmq` 2.0 nebude fungovat s Nette 2.4.

V Nette 3 se například používají v interfacech metody s otypovaným výstupem, což pak způsobuje problémy, protože pokud se to tak nenapíše do implementace interface, tak nesedí signatura a stejně tak když to je v implementaci navíc. [Viz tady.](https://github.com/peldax/RabbitMq/compare/v1.2.0...AppsDevTeam:master#diff-c68f8b39dc1e83fd3967f4b52fd4fcba)

NoxZet committed:
Different registering of shutdown listener based on Kdyby\Events

- If Kdyby\Events\SymfonyDispatcher is present, use Kdyby EventManager
- Else use implementation of Symfony EventDispatcherInterface

NoxZet committed:
Update composer.json, remove Nette 2.3, replace rabbitmq

thorewi commented on src/DI/BackgroundQueueExtension.php:
```diff
@@ -101,17 +102,27 @@ public function afterCompile(ClassType $class)
-			$service->onShutdown();
-		});
+	
+	if (get_class($this->getByType(\Symfony\Component\EventDispatcher\EventDispatcherInterface::class)) === \Kdyby\Events\SymfonyDispatcher::class) {
```
proc tady neni instanceof?

NoxZet replied:
Šlo tady o ošetření toho, že daná třída `SymfonyDispatcher` postrádá implementaci addListener. Myslím, že intanceof pak nemá moc smysl, protože pokud by někdo chtěl tuhle třídu extendnout a doimplementovat si metodu addListener, pak to bude tuhle novou třídu pořád ignorovat a brát místo ní dále `EventManager`.

NoxZet committed:
Remove Kdyby SymfonyDispatcher and EventManager support

NoxZet committed:
Dependencies compatible with both Nette 2.3+ and Nette 3.0.
